File: Stasis.Engine/UI/UDim.cs
using Raylib_cs;

namespace Stasis.Engine.UI;

public struct UDim {
    public float Scale { get; set; }
    public float Offset { get; set; }

    public static UDim Zero => new UDim(0, 0);

    public UDim(float scale, float offset) {
        Scale = scale;
        Offset = offset;
    }
}

public struct UDim2 {
    public UDim X;
    public UDim Y;

    public static UDim2 Zero => new UDim2(0, 0, 0, 0);
    public static UDim2 Fill => new UDim2(1, 0, 1, 0);

    public UDim2(UDim x, UDim y) {
        X = x;
        Y = y;
    }
    public UDim2(float xScale, float xOffset, float yScale, float yOffset) {
        X = new UDim(xScale, xOffset);
        Y = new UDim(yScale, yOffset);
    }

    public static UDim2 ScaleByPixels(float xPixels, float yPixels) {
        return new UDim2(xPixels / Raylib.GetRenderWidth(), 0, yPixels / Raylib.GetRenderHeight(), 0f);
    }
}

File: src/core/ui/udim.rs
pub struct UDim {
    pub scale: f32,
    pub offset: f32,
}

impl UDim {
    pub fn new(scale: f32, offset: f32) -> Self {
        Self {
            scale, offset
        }
    }

    pub fn zero() -> Self {
        Self {
            scale: 0.,
            offset: 0.,
        }
    } 
}

pub struct UDim2 {
    pub x: UDim,
    pub y: UDim,
}

impl UDim2 {
    pub fn new(x_scale: f32, x_offset: f32, y_scale: f32, y_offset: f32) -> Self {
        Self {
            x: UDim::new(x_scale, x_offset),
            y: UDim::new(y_scale, y_offset),
        }
    }

    pub fn zero() -> Self {
        Self {
            x: UDim::zero(),
            y: UDim::zero(),
        }
    }

    pub fn fill() -> Self {
        Self::new(1., 0., 1., 0.)
    }
}